src/runtime_src/core/tools/xbtracer/src/lib/capture.cpp

```diff
@@ -0,0 +1,617 @@
+  if (status == 0)
+  {
+    // Use std::unique_ptr with a custom deleter to manage demangled_name
+    std::unique_ptr<char, decltype(&std::free)> demangled_ptr(demangled_name,
+                                                              std::free);
+
+    /* Conditioning of demangled name because of some platform based deviation
+      */
+    std::vector<std::pair<std::string, std::string>> replacements =
+      {
+        {"std::__cxx11::basic_string<char, std::char_traits<char>, "
+              "std::allocator<char> >", "std::string"},
+        {"[abi:cxx11]", ""},
+        {"std::map<std::string, unsigned int, std::less<std::string >, "
+            "std::allocator<std::pair<std::string const, unsigned int> > > "
+            "const&", "xrt::hw_context::cfg_param_type const&"}
+      };
+
+    std::string result =
+        find_and_replace_all(std::string(demangled_name), replacements);
+
+    return result;
+  }
+  else
+    // Demangling failed
+    return {mangled_name};  // Return the original mangled name
+}
+
+static std::string find_library_path()
+{
+  std::lock_guard<std::mutex> lock(env_mutex);
+  // NOLINTNEXTLINE(concurrency-mt-unsafe) - protected by env_mutex
+  char* ld_preload = getenv("LD_PRELOAD");
+  if (ld_preload == nullptr)
+  {
+    std::cout << "LD_PRELOAD is not set." << std::endl;
+    return "";
+  }
+
+  std::string full_path = std::string(ld_preload);
+  full_path.erase(std::remove(full_path.begin(), full_path.end(), ' '),
+                  full_path.end());
+  return full_path;
+}
+
+/**
+ * This function will update the dispatch table
+ * with address of the functions from original
+ * library.
+ */
+int router::load_func_addr()
+{
+  // Load the shared object file
+  handle = dlopen(lib_name, RTLD_LAZY);
+  if (!handle)
+  {
+    std::lock_guard<std::mutex> lock(dlerror_mutex);
+    // NOLINTNEXTLINE(concurrency-mt-unsafe) - protected by dlerror_mutex
+    std::cerr << "Error loading shared library: " << dlerror() << std::endl;
+    return 0;
+  }
+
+  /**
+   * Get Function address's which are of interest and ignore others.
+   */
+  for (auto& it : func_mangled)
+  {
+    auto ptr_itr = fname2fptr_map.find(it.first);
+
+    if (ptr_itr != fname2fptr_map.end())
+    {
+      void** temp = ptr_itr->second;
+      /* update the original function address in the dispatch table */
+      *temp = (dlsym(handle, it.second.c_str()));
+      if (nullptr == temp)
+        std::cout << "Null Func address received " << std::endl;
+      else
+        fptr2fname_map[*temp] = std::regex_replace(it.first,
+                                      std::regex(R"(\)\s*const)"), ")");
+    }
+    else
+    {
+      // std::cout << "func :: \"" << it.first << "\" not found in"
+      //          << "fname2fptr_map\n";
+    }
+  }
+  return 1;
+}
+
+/**
+ * This function will read mangled API's from library and perform
+ * Demangling operation.
+ */
+int router::load_symbols()
```

Comment:
ditto on return error code here.

Also this is quite a long functions, any possibility of refactoring to make it more readable?

Reply:
These comments were folded and overlooked them, my bad. I have address these, please check.